pkg/wakaama/contrib/objects/light_control.c

```diff
@@ -0,0 +1,614 @@
+}
+
+static uint8_t _read_cb(uint16_t instance_id, int *num_data, lwm2m_data_t **data_array,
+                        lwm2m_object_t *object)
+{
+    lwm2m_obj_light_control_inst_t *instance;
+    uint8_t result = COAP_404_NOT_FOUND;
+    int i = 0;
+
+    mutex_lock(&_light_control_object.lock);
+
+    /* try to get the requested instance from the object list */
+    instance = (lwm2m_obj_light_control_inst_t *)lwm2m_list_find(object->instanceList, instance_id);
+    if (!instance) {
+        DEBUG("[lwm2m:light_control:read]: can't find instance %d\n", instance_id);
+        result = COAP_404_NOT_FOUND;
+        goto free_out;
+    }
+
+    /* if the number of resources is not specified, we need to read all resources */
+    if (!*num_data) {
+        DEBUG("[lwm2m:light_control:read]: reading all resources\n");
+
+        uint16_t res_list[] = {
+            LWM2M_LIGHT_CONTROL_ON_OFF_ID,
+            LWM2M_LIGHT_CONTROL_DIMMER_ID,
+            LWM2M_LIGHT_CONTROL_ON_TIME_ID,
+            LWM2M_LIGHT_CONTROL_COLOUR_ID,
+            LWM2M_LIGHT_CONTROL_APP_TYPE_ID
+        };
+
+        /* allocate structures to return resources */
+        int res_num = ARRAY_SIZE(res_list);
+        *data_array = lwm2m_data_new(res_num);
+
+        if (NULL == *data_array) {
+            result = COAP_500_INTERNAL_SERVER_ERROR;
+            goto free_out;
+        }
+
+        /* return the number of resources being read */
+        *num_data = res_num;
+
+        /* set the IDs of the resources in the data structures */
+        for (i = 0; i < res_num; i++) {
+            (*data_array)[i].id = res_list[i];
+        }
+    }
+
+    /* now get the values */
+    i = 0;
+    do {
+        DEBUG("[lwm2m:light_control:read]: reading resource %d\n", (*data_array)[i].id);
+        result = _get_value(&(*data_array)[i], instance);
+        i++;
+    } while (i < *num_data && COAP_205_CONTENT == result);
+
+free_out:
+    mutex_unlock(&_light_control_object.lock);
+    return result;
+}
+
+static uint8_t _write_cb(uint16_t instance_id, int num_data, lwm2m_data_t * data_array,
+                         lwm2m_object_t * object)
+{
+    lwm2m_obj_light_control_inst_t *instance;
+    uint8_t result = COAP_204_CHANGED;
+    bool call_cb = false;
+
+    mutex_lock(&_light_control_object.lock);
+
+    /* try to get the requested instance from the object list */
+    instance = (lwm2m_obj_light_control_inst_t *)lwm2m_list_find(object->instanceList, instance_id);
+    if (!instance) {
+        DEBUG("[lwm2m:light_control:write]: can't find instance %d\n", instance_id);
+        result = COAP_404_NOT_FOUND;
+        goto free_out;
+    }
+
+    for (int i = 0; i < num_data && result == COAP_204_CHANGED; i++) {
+        switch (data_array[i].id) {
+        case LWM2M_LIGHT_CONTROL_ON_OFF_ID: {
+            bool prev_status = instance->status;
+
+            lwm2m_data_decode_bool(&data_array[i], &instance->status);
+
+            if (instance->status && !prev_status) {
```

Comment:
I don´t understand what `prev_status` accomplishes in this `case`?

Reply:
`prev_status` holds the status of the light previous to the change by `lwm2m_data_decode_bool`. It's used so that the stopwatch is only reset on OFF -> ON transitions.

Reply:
It should only count the On time, but now that I think about it, I should stop and reset it whenever an OFF is written...

Reply:
Ok so because of the way `ztimer_stopwatch` works, I can't stop it because it releases it, then I can't reset it. I chose to keep it running always and resting the count when OFF -> ON transition occurs. When the timer count is read, if the light is OFF then the count is 0.